File: java/PortScanner/PortScan.java
import java.io.IOException;
import java.net.InetSocketAddress;
import java.net.Socket;

// Run with:
// javac PortScan.java
// java PortScan 127.0.0.1

public class PortScan {

    private static void printUsage() {
        System.out.println("Usage: ");
        System.out.println(" java PortScan <host>");
        System.out.println("Examples: ");
        System.out.println(" java PortScan 127.0.0.1");
    }

    private static void checkPort(String host, int portNumber) {
        Thread thread = new Thread(() -> {
            Socket clientSocket = new Socket();
            try {
                clientSocket.connect(new InetSocketAddress(host, portNumber), 3000);
                System.out.printf("[+] %d Connected\n", portNumber);
                clientSocket.close();
            } catch (IOException e) {
                //System.out.printf("[-] Could not connect %d");
                //e.printStackTrace();
            }
        });
        thread.start();
    }

    public static void main(String[] args) {

        if (args.length == 0) {
            System.out.println("No arguments received.");
            printUsage();
            System.exit(1);
        }

        String host = args[0];
        // System.out.println(args);
        // System.out.println(host);
        System.out.printf("Scanning %s now...", host);

        // port 1-65535
        // Open a TCP connection for each port
        //

        for (Integer portNumber = 1; portNumber <= 65535; portNumber++) {
            checkPort(host, portNumber);
        }

    }

}

File: go/portscan/portscan.go
package main

//Run with:
// go run portscan.go 127.0.0.1

import (
	"fmt"
	"log"
	"net"
	"os"
	"strconv"
	"time"
)

func printUsage() {
	fmt.Println("Usage: ")
	fmt.Println("	go run portscan.go <host>")
	fmt.Println("Example: ")
	fmt.Println("	go run portscan.go 127.0.0.1")
}

func testTcpConnect(host string, port int, doneChannel chan bool) {
	timeoutLength := 5 * time.Second
	conn, err := net.DialTimeout("tcp", host+":"+strconv.Itoa(port), timeoutLength)
	if err != nil {
		doneChannel <- false
		return // could not connect
	}
	conn.Close()
	log.Printf("[+] %d connected", port)
	doneChannel <- true
}

func main() {
	if len(os.Args) == 1 {
		log.Println("No arguments provided.")
		printUsage()
		os.Exit(1)
	}

	doneChannel := make(chan bool)
	activeThreadCount := 0
	log.Println("Scanning host: " + os.Args[1])
	for portNumber := 1; portNumber <= 65535; portNumber++ {
		activeThreadCount++
		go testTcpConnect(os.Args[1], portNumber, doneChannel)
	}

	for {
		<-doneChannel
		activeThreadCount--
		if activeThreadCount == 0 {
			break
		}

	}
	// until activeThreadCount == 0 keep checking
	//

}
